test/test_pty.rs

```diff
@@ -90,3 +93,81 @@ fn test_open_ptty_pair() {
     let slave_fd = open(Path::new(&slave_name), O_RDWR, stat::Mode::empty()).unwrap();
     assert!(slave_fd > 0);
 }
+
+#[test]
+fn test_openpty() {
+    let pty = openpty(None, None).unwrap();
+    assert!(pty.master > 0);
+    assert!(pty.slave > 0);
+
+    // writing to one should be readable on the other one
+    let string = "foofoofoo\n";
+    let mut buf = [0u8; 16];
+    write(pty.master, string.as_bytes()).unwrap();
+    let len = read(pty.slave, &mut buf).unwrap();
+
+    assert_eq!(len, string.len());
+    assert_eq!(&buf[0..len], string.as_bytes());
+
+    // read the echo as well
+    let echoed_string = "foofoofoo\r\n";
+    let len = read(pty.master, &mut buf).unwrap();
+    assert_eq!(len, echoed_string.len());
+    assert_eq!(&buf[0..len], echoed_string.as_bytes());
+
+    let string2 = "barbarbarbar\n";
+    let echoed_string2 = "barbarbarbar\r\n";
+    write(pty.slave, string2.as_bytes()).unwrap();
+    let len = read(pty.master, &mut buf).unwrap();
+
+    assert_eq!(len, echoed_string2.len());
+    assert_eq!(&buf[0..len], echoed_string2.as_bytes());
+
+    close(pty.master).unwrap();
+    close(pty.slave).unwrap();
+}
+
+#[test]
+fn test_openpty_with_termios() {
+    // open one pty to get attributes for the second one
+    let mut termios = {
+        let pty = openpty(None, None).unwrap();
+        assert!(pty.master > 0);
+        assert!(pty.slave > 0);
+        let termios = tcgetattr(pty.master).unwrap();
+        close(pty.master).unwrap();
+        close(pty.slave).unwrap();
+        termios
+    };
+    termios.c_oflag &= !ONLCR;
+
+    let pty = openpty(None, &termios).unwrap();
+    assert!(pty.master > 0); // must be valid file descriptors
+    assert!(pty.slave > 0);
+
+    // writing to one should be readable on the other one
+    let string = "foofoofoo\n";
+    let mut buf = [0u8; 16];
+    write(pty.master, string.as_bytes()).unwrap();
+    let len = read(pty.slave, &mut buf).unwrap();
+
+    assert_eq!(len, string.len());
+    assert_eq!(&buf[0..len], string.as_bytes());
+
+    // read the echo as well
+    let echoed_string = "foofoofoo\n";
+    let len = read(pty.master, &mut buf).unwrap();
+    assert_eq!(len, echoed_string.len());
+    assert_eq!(&buf[0..len], echoed_string.as_bytes());
+
+    let string2 = "barbarbarbar\n";
+    let echoed_string2 = "barbarbarbar\n";
+    write(pty.slave, string2.as_bytes()).unwrap();
+    let len = read(pty.master, &mut buf).unwrap();
+
+    assert_eq!(len, echoed_string2.len());
+    assert_eq!(&buf[0..len], echoed_string2.as_bytes());
+
+    close(pty.master).unwrap();
```

Comment:
I'm not sure; the best documentation I could find seemed to imply either way was okay (although nothing implies that closing either end will close the other one):

```
If fildes refers to the master side of a pseudo-terminal, and this is the last close, a SIGHUP signal shall be sent to the controlling process, if any, for which the slave side of the pseudo-terminal is the controlling terminal. It is unspecified whether closing the master side of the pseudo-terminal flushes all queued input and output.

If fildes refers to the slave side of a STREAMS-based pseudo-terminal, a zero-length message may be sent to the master.

```